common/lib/dependabot/file_fetchers/base.rb

```diff
@@ -91,7 +95,34 @@ def fetch_file_if_present(filename, fetch_submodules: false)
         raise Dependabot::DependencyFileNotFound, path
       end
 
+      def load_cloned_file_if_present(filename)
+        path = File.join(clone_repo_contents, filename)
+        unless File.exist?(path)
+          raise Dependabot::DependencyFileNotFound, filename
+        end
+
+        content = File.read(path)
+        type = if File.symlink?(path)
+                 symlink_target = File.readlink(path)
+                 "symlink"
+               else
+                 "file"
+               end
+
+        DependencyFile.new(
+          name: Pathname.new(filename).cleanpath.to_path,
+          directory: directory,
+          type: type,
+          content: content,
+          symlink_target: symlink_target
+        )
+      end
+
       def fetch_file_from_host(filename, type: "file", fetch_submodules: false)
+        unless @repo_contents_path.nil?
+          return load_cloned_file_if_present(filename)
+        end
+
         path = Pathname.new(File.join(directory, filename)).cleanpath.to_path
         content = _fetch_file_content(path, fetch_submodules: fetch_submodules)
         type = @linked_paths.key?(path.gsub(%r{^/}, "")) ? "symlink" : type
```

Comment:
Looks like `repo_contents` is also used from go_modules and other places to list a directories contents

Reply:
Can you expand on the implications of this?

Reply:
Oh the existing error case for `repo_contents` is really hard to parse. Submodule handing stands out as a painful thing, do we need to start calling something like `git submodule update --init --recursive` to get all the submodule content checked out?

Reply:
> Can you expand on the implications of this?

Should we `ls` the contents of the folder in the clone instead of fetching from github?

Reply:
> Submodule handing stands out as a painful thing, do we need to start calling something like `git submodule update --init --recursive` to get all the submodule content checked out?

Looks like we don't fetch submodules from any of the go modules so maybe not worth implementing that now, would be good to add a comment explaining this functionality is missing from `load_cloned_file_if_present`.

Reply:
> would be good to add a comment explaining this functionality is missing from load_cloned_file_if_present

I aimed for that in https://github.com/dependabot/dependabot-core/pull/2578/commits/a98c4ddfe65ec25f82fe08990272cef612a1f6d8#diff-a619536a45c6fc217c2f5e4c56460121R35-R43 - note the limitations and punt.

I think the way [`Dir.glob`](https://github.com/dependabot/dependabot-core/pull/2551/files#diff-d76f068189ee62d3127d9dbc19dcf4d0R74) is used in the eventual downstream where this is to be enabled, the `GoModules::FileFetcher` can stop calling `repo_contents` as soon as it can `repo_contents_path.nil?` to know the new glob path will work.

Reply:
> GoModules::FileFetcher can stop calling repo_contents as soon as it can repo_contents_path.nil? to know the new glob path will work.

Gotcha, that's way nicer!